base/terminology_processor.bal

```diff
@@ -14,206 +14,1437 @@
+
+            codeSystems = from CodeSystem entry in codeSystems
+                where entry.id == id
+                select entry;
+
+            if codeSystems.length() > 0 {
+                if 'version is string {
+                    codeSystems = from CodeSystem entry in codeSystems
+                        where entry.'version == 'version
+                        select entry;
+                    cs = codeSystems.length() > 0 ? codeSystems[0] : cs;
+                }
+                if codeSystems.length() < 1 {
+                    return createFHIRError(
+                        string `Unknown version: '${'version.toString()}'`,
+                        ERROR,
+                        PROCESSING_NOT_FOUND,
+                        diagnostic = string
+                        `There is CodeSystem in the registry with Id: '${id}' but can not find version: '${'version.toString()}' of it`,
+                        httpStatusCode = http:STATUS_NOT_FOUND
+                        );
+                } else {
+                    string latestVersion = "0.0.0";
+                    foreach var item in codeSystems {
+                        if item.'version > latestVersion {
+                            latestVersion = <string>item.'version;
+                            cs = item;
+                        }
+                    }
+                }
+            } else {
+                return createFHIRError(
+                    string `Unknown CodeSystem: '${id}'`,
+                                                ERROR,
+                                                PROCESSING_NOT_FOUND,
+                                                httpStatusCode = http:STATUS_NOT_FOUND
+                                                );
+            }
+
+            return cs.clone();
+        }
+    }
+
+    # Find a ValueSet for a provided Id and version.
+    #
+    # + id - Id of the Value Set to be retrieved
+    # + version - Version of the ValueSet to be retrieved and it should be provided with system parameter,
+    # if this version parameter is not supplied then the latest version of CodeSystem will picked up.
+    # + return - Return ValueSet data if the request is successful, return FHIR error if no data found for the provided Id
+    public isolated function readValueSetById(string id, string? 'version = ()) returns ValueSet|FHIRError {
+        lock {
+            ValueSet[] valueSets = self.valueSets.toArray();
+            ValueSet vs = {status: "unknown"};
+
+            valueSets = from ValueSet entry in valueSets
+                where entry.id == id
+                select entry;
+
+            if valueSets.length() > 0 {
+                if 'version is string {
+                    valueSets = from ValueSet entry in valueSets
+                        where entry.'version == 'version
+                        select entry;
+                    vs = valueSets.length() > 0 ? valueSets[0] : vs;
```

Comment:
Let's have if block to set `valueSets[0]` to `vs`, it is not needed to assign `vs` to `vs`

Reply:
Done